### ISSUE TITLE
Optimize the stuff around the consistent sampler

### ISSUE THREAD

**Issue by mcchilders** [votingworks/arlo#962]
We're running into instances where the consistent sampler implementation is suuuuuuuper slow/resource intensive upwards of 6M ballots ... let's do that thing a little bit better!

**Comment by cjerdonek** [comment]
I'm not sure how and if you're actually using the "consistency" aspect of the sampling. But either way, I'm pretty sure there's a relatively straightforward way to do consistent sampling with  `2 * take` hashes instead of [`len(id_list)` + `take` hashes](https://github.com/ron-rivest/consistent_sampler#efficiency), if you know the batches you want to be consistent with ahead of time. (If you don't need the consistency, I think `take` hashes would suffice.) That would mean e.g. you wouldn't need to do 6M hashes before starting.

**Comment by jonahkagan** [comment]
@cjerdonek My understanding is that we use the "consistency" to reduce the amount of ballots that need to be audited in the case that the audit is targeting multiple contests in the same state, which each have their own (likely overlapping) set of ballots. We have to draw a sample from the set of ballots for each contest and then union the samples. Consistency lowers the size of the resulting union. (@mcchilders, @umbernhard, or @benadida can correct me if I'm wrong)

What do you mean by "know the batches you want to be consistent with ahead of time"? Does the situation I outlined above satisfy that requirement?

Would love to hear the method you have for consistent sampling with less hashes!

**Comment by cjerdonek** [comment]
@jonahkagan Thanks, Jonah. By "know the batches you want to be consistent with ahead of time," what I was getting at is whether the subsets you want consistency for are fixed subsets you know in advance -- things like counties, contests, etc. (Based on your reply, it sounds like that's the case for you.) The reason I ask is that the method in [Ron Rivest's paper](https://arxiv.org/abs/1808.10016) has the consistency property for arbitrary subsets, which seems like is stronger than you need.

I have a couple more follow-up questions that hopefully you can answer:

1. In [Ron's README](https://github.com/ron-rivest/consistent_sampler), he writes:

   > For our applications, one big advantage of consistent sampling is the following. If each county collects cast ballots separately, then they can order their own ballots for sampling and interpretation independently of what other counties are doing. An overall sample order can be constructed from the individual county sample order, by merging the list of triples each produces into an overall sorted order.
  
   In your reply, you said you needed consistency for contests, which are the subsets the ballots are getting sampled _from_. So my question is whether you also use the consistency Ron is describing above, which is more about how the ballots are _stored and organized_, independent of contest (e.g. by county).
  
2. When you're targeting multiple contests, do you compute the samples for all contests all at once in a single operation? Or are the samples computed for contests individually, e.g. requiring a separate user interaction for each?

**Comment by nealmcb** [comment]
I think the goal should be the ability to handle arbitrary subsets, as Ron outlines.
And the subsets can be by ballot organization in space (counties, overlapping districts etc) or in time (election-day ballots, provisional ballots counted later, etc).  There would of course need to be a separate random draw in any case where the manifest was updated after the previous random draw.

**Comment by jonahkagan** [comment]
@cjerdonek let me see if I can answer those questions. I think an explanation of how we do sampling in Arlo might answer both.

The sampling algorithm in Arlo goes something like this:
```python
# Draw a sample for each contest, then union it into the overall set of sampled ballots
sampled_ballots = set()
for each contest:
   # Create a disjoint union of ballots across counties to sample from
  ballot_pool = list()
  for each county:
    tagged_county_ballots = [(county.name, ballot.id) for ballot in county.ballots]
    ballot_pool += tagged_county_ballots
  # Draw a sample from that pool
  sampled_ballots += consistent_sampler.draw_sample(ballot_pool)
return sampled_ballots
```

So I think that means the answers to your questions are:
1. We do merge the ballot orderings from different counties, but we do it before sampling, tagging the ballots with the name of the county they came from. We're not drawing samples on a county by county basis and then combining the resulting samples.
2. All of the sampling happens in a single interaction from the user's perspective, but behind the scenes we are drawing a separate sample for each contest and then unioning them.

**Comment by cjerdonek** [comment]
@jonahkagan Thanks for the explanation. I'll see if I can explain the approach I have in mind briefly. If something's not clear, feel free to ask for clarification.

At its core, the approach uses the central idea of Ron Rivest's "pseudo-random sampler" procedure: https://people.csail.mit.edu/rivest/sampler.py
That procedure has as its basic building block a function that can be used to pick a "random" number between 1 and an integer N, given seed data. It does this using a hash followed by a modulo operation (so no floats needed). In the below, I'll refer to using this function once as "sampling."

Given e.g. a state, the idea is to do selections in the state using a "layered" approach. Any number of layers can be used, but I'll describe it using two layers (using the layering defined by counties). Say you're auditing a statewide contest with N ballots and K counties. The N ballots are divided into K buckets by the counties. To pick your first ballot, first "sample" from N. That gives you a bucket (i.e. one of the K counties). (Here you don't care about which ballot in the bucket it is.) Note that the likelihood of a county being selected using this process is proportional to how many ballots it has, which is what you want. Say the county you picked has M ballots. Now "sample" from M to get the ballot. So with two hashes you got your first ballot. Then, you keep doing this, first sampling from the state to get a bucket, and then sampling from within a bucket to get a ballot, until you've sampled enough ballots. Each ballot selection requires two hashes. This method really excels for statewide contests because the number of hashes is only twice the number of ballots you need to pick, and you don't need to do any sorting or float arithmetic either.

If you need to do a smaller contest, say one inside a county, then you can skip the first step and just "sample" from M. This selection will be consistent with the statewide selection because the statewide selection used the county ordering when it came to selecting from within a county.

There are other details I haven't mentioned, but that's the basic idea. Does that give you enough to reconstruct what I had in mind?

**Comment by jonahkagan** [comment]
Yes, that definitely gives a clear idea of the sampling method. We've talked it through a bit and have a few questions that we'd need answered in order to convince ourselves that this would meet the requirements for Arlo:
1. Does this method produce a uniform random sample? Will it sample ballots from both state-level and county level contest uniformly?
2. Does this method work with replacement?
3. Will this method maximize the overlap between samples drawn for different counties?

To explain question 3 in a bit more detail: Each contest in Arlo is associated with a set of counties that had that contest on their ballots - we call this the "contest universe." When we draw the sample for each contest, the pool we sample from is the union of the ballots from each county in the contest universe. In other words, in the pseudocode I wrote above, the line `for each county:` should actually be `for each county in contest.counties`. The consistent sampling algorithm that we use today ensures that there will be maximum possible overlap between the samples drawn for each contest, which helps reduce the actual real-world work of retrieving and auditing ballots

We'd like to hear your thoughts on these questions.

Also, just for your info, for core math changes like this, we need to make sure that we get it reviewed by a wider group in the RLA community. We can work with you to do that if we're all feeling confident that this method has potential. We'd probably need some sort of written proposal and/or code sample in order to move forward with that.

**Comment by cjerdonek** [comment]
Thanks for considering, Jonah (and others).

> In other words, in the pseudocode I wrote above, the line `for each county:` should actually be `for each county in contest.counties`.

Yeah, I figured that. Thanks for clarifying.

As to your questions--

> 1. Does this method produce a uniform random sample? Will it sample ballots from both state-level and county level contest uniformly?

Yes, it's uniform, in part for the same reason that Rivest's "pseudo-random sampler" is uniform (namely that SHA-256 is pseudo-random). The other reason it's uniform is that when choosing "buckets" in a layer by choosing an integer between 1 and N, you always update N and the bucket sizes before choosing. That ensures that the likelihood of choosing a bucket is always exactly proportional to the number of ballots in the bucket. (You only need to do these updates in the case of sampling without replacement, because only in that case does the total number of ballots change as you move forward through the process.)

> 2. Does this method work with replacement?

Yes, that's actually the easier case, because you don't need to do the totals updates that I was referring to above. The ballots and buckets you're choosing from are the same each time.

> 3. Will this method maximize the overlap between samples drawn for different counties?

Yes, the approach has the same union / consistency property as your current method in this regard. The reason is that the method I described can be viewed as providing a "global" sampling order, in the same way that the hash-to-float-then-sort method that you're currently using provides a global ordering. What's different about the method I described is that you can build up the ordering from the beginning, instead of having to compute the full ordering for every ballot before knowing the first item. The method I described also has the property that if your contest lies entirely within one "layer," then you can dispense with stages that involve higher layers and skip to working with the smallest applicable layer.

The one trade-off in the method is what I'll describe right now. If your "ballot universe" lives at one layer, but it doesn't encompass all buckets in the layer, then you still need to sample from the full layer, but at the cost of having to re-sample if the ballot you picked doesn't lie within the contest. To use your example, if you have two layers, the full state and individual counties, and you're auditing a contest that involves multiple counties (but not the full state), then when you're doing the first stage of choosing a bucket (aka a county), then you'll need to re-sample each time you choose a bucket or ballot that doesn't overlap with the contest.

You have a lot of flexibility in this regard though that you can prepare for in advance. For example, if you know your multi-county contests are congressional races that cross multiple counties, then you can make the layers be the full state and the congressional districts instead of the full state and counties. (This is also what I was getting at above when I first asked "if you know the batches you want to be consistent with ahead of time.") Or you can stick with counties, at the expense of having to do some number of extra hashes because of sampling ballots outside of the contest. Does that make sense?

> Also, just for your info, for core math changes like this, we need to make sure that we get it reviewed by a wider group in the RLA community.

Yeah, totally. I'm happy to help with that where I can if you're interested in exploring this approach.

**Comment by cjerdonek** [comment]
By the way, something else I haven't mentioned is that I think there may also be opportunities for improvement by doing a single sampling process and iterating over all contests after each choice of ballot, instead of doing a separate full sampling process for each contest, and then taking a union.

With this approach, you would pick your first ballot globally, and then ask, "which contests does this apply to"? And do this with each ballot. As you move through the process and certain contests are fulfilled earlier than others, you can start to skip over areas until you are left, generally speaking, with the contests with the tightest margins towards the end.

**Comment by nealmcb** [comment]
Can someone summarize the impact of a "wontfix" here?

- Is it still "suuuuuuuper slow/resource intensive upwards of 6M ballots"?
- Can Arlo draw a state-wide sample for e.g. California? Does it require any particular hardware to be usable?
- Do you think one of Chris' ideas would be worthwhile if someone else offered a PR?
